### PR TITLE
feat(wip): Support for pre and post runner run hooks

### PR DESCRIPTION
**Pull request by eliorc**
## Description
This PR introduces the ability to attach hooks for pre and post `runner.run` calls.
It introduces methods for attaching hooks, e.g. `runner.add_post_run_hook(...)`, `runner.remove_pre_run_hook(...)` and makes sure these hooks are executed on the different `.run` flavors, async, batch, etc.

## Motivation and Context
Motivation is detailed in #2471 

## How Has This Been Tested?

Tested only with the already-existing tests

## Checklist:
- [ ] My code follows the bentoml code style, both `make format` and
  `make lint` script have passed
  ([instructions](https://github.com/bentoml/BentoML/blob/master/DEVELOPMENT.md#style-check-and-auto-formatting-your-code)).  - *Partially, I ran the make commands, but some show errors, will need help on these*
- [] My change reduces project test coverage and requires unit tests to be added
- [ ] I have added unit tests covering my code change
- [x] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

**Comment by eliorc** [comment]
I kept this as a draft PR to get some feedbacks from the maintainers, would love to hear your thoughts

**Comment by parano** [comment]
Hi @eliorc, thanks for the suggestion! I really like the idea of customizing the model runner.

Our team is currently working on a fairly big redesign of the Runner internal API in this branch here: https://github.com/bentoml/BentoML/tree/runners-1.0 - this refactoring will make it easier for users to define custom runners, and control runner resources and provisioning behaviors. The development branch has changed lots of code around the runner code path, maybe after the runner-1.0 branch get merged we can revisit this again and build it on top of the new design?

**Comment by eliorc** [comment]
@parano 
Thanks for the heads up, this is exactly why I opened the draft to make sure I won't code in vain.
So when do we expect the new runners to be out? The monitoring thing is really important to our organization and I'm sure for others as well

@aarnphm will change to Google-style once I can implement this in the relevant branch

File: bentoml/_internal/runner/runner.py [hunk 1]
```diff
@@ -152,21 +202,70 @@ def batch_options(self) -> BatchOptions:
...
+        """
+
+        for hook in self._post_run_hooks.values():
+            hook(*args, **kwargs)
+
     @final
     async def async_run(self, *args: t.Any, **kwargs: t.Any) -> t.Any:
-        return await self._impl.async_run(*args, **kwargs)
+
+        self._execute_pre_run_hooks(*args, **kwargs)
```
Comment (aarnphm) [inline bentoml/_internal/runner/runner.py:226]: quick thoughts: It would be nice if we can execute pre_run hook in an async loop? 💭

Comment (eliorc) [reply bentoml/_internal/runner/runner.py:226]: Indeed, though I think its irrelevant to discuss such implementation details until we understand how implementing this feature in the new runner's implementation architecture is possible

**Comment by aarnphm** [comment]
Hi @eliorc, since the RC is just released, we are more than happy to chat with you about this feature :) Feel free to poke around the new code base 😃

**Comment by eliorc** [comment]
I'll do that and update! Might take a while though, the day after tomorrow I'm going for a month's vacation

**Comment by parano** [comment]
@aarnphm we probably don't need the runner hook anymore, with the introduction of custom runnable: https://docs.bentoml.org/en/latest/concepts/runner.html#custom-runner

cc @eliorc

**Pull request closed without merging**